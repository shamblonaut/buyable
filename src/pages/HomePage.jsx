import { Link, useOutletContext } from "react-router-dom";
import styled from "styled-components";

import Carousel from "../components/Carousel";

const Page = styled.main`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 16px;
  flex: 1;
`;

const HeroSection = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
`;

const HeroText = styled.p`
  font-size: 2rem;
  font-weight: 500;
  text-align: center;
  margin-top: 32px;
`;

const CTAButton = styled(Link)`
  margin-top: 32px;
  cursor: pointer;

  & button {
    padding: 8px 16px;
    font-size: 1.25rem;
    font-weight: 500;
  }
`;

const Loading = styled.p`
  margin: 32px;
`;

const HomePage = () => {
  const { productsData, productsLoading } = useOutletContext();

  return (
    <Page>
      <HeroSection>
        {productsLoading ? (
          <Loading>Loading Carousel...</Loading>
        ) : (
          productsData && <Carousel items={productsData.slice(0, 11)} />
        )}
        <HeroText>Buy to your heart's content!</HeroText>
        <CTAButton to="/shop">
          <button>Explore Products</button>
        </CTAButton>
      </HeroSection>
    </Page>
  );
};

export default HomePage;
