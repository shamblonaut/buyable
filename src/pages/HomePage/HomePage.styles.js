import { Link } from "react-router-dom";
import styled from "styled-components";

export const Page = styled.main`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 16px;
  flex: 1;
`;

export const HeroSection = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
`;

export const HeroText = styled.p`
  font-size: 2rem;
  font-weight: 500;
  text-align: center;
  margin-top: 32px;
`;

export const CTAButton = styled(Link)`
  margin-top: 32px;
  cursor: pointer;

  & button {
    padding: 8px 16px;
    font-size: 1.25rem;
    font-weight: 500;
  }
`;

export const Loading = styled.p`
  margin: 32px;
`;
