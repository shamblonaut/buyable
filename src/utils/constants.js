export const AppPosition = Object.freeze({
  HOME: 0,
  SHOP: 1,
  CART: 2,
});
