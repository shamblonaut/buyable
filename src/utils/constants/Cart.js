export const CartAction = Object.freeze({
  ADDED: "added_to_cart",
  REMOVED: "removed_from_cart",
  INCREMENTED: "incremented_cart_quantity",
  DECREMENTED: "decremented_cart_quantity",
  EDITED: "edited_cart_quantity",
  CLEARED: "cleared_cart",
});
