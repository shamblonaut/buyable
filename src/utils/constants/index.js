import { AppPosition } from "./App";

export { AppPosition };
