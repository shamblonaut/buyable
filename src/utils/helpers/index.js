import { getPositionRoute } from "./App";
import { getRatingArray } from "./Rating";
import { toTitleCase } from "./Misc";

export { getPositionRoute, getRatingArray, toTitleCase };
