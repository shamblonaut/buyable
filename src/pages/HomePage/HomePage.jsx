import { useEffect } from "react";
import { useOutletContext } from "react-router-dom";

import { AppPosition } from "@/utils/constants";

import { Carousel } from "@/components";
import { Page } from "@/styles";
import {
  HeroSection,
  HeroText,
  HeroHighlight,
  CTALink,
} from "./HomePage.styles";

const HomePage = () => {
  const { productsData, setAppPosition } = useOutletContext();

  useEffect(() => setAppPosition(AppPosition.HOME), [setAppPosition]);

  return (
    <Page>
      <Carousel productsData={productsData} itemCount={10} />
      <HeroSection>
        <HeroText>
          I can't believe it's not
          <br />
          <HeroHighlight>BUYABLE</HeroHighlight>
        </HeroText>
        <CTALink to="/shop" aria-label="Go to Shop page">
          Explore Products
        </CTALink>
      </HeroSection>
    </Page>
  );
};

export default HomePage;
