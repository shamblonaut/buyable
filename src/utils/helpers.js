import { AppPosition } from "./constants";

export const getRatingArray = (rating) => {
  if (rating > 5.0) {
    throw new Error("Invalid rating");
  }

  let ratingArray = [];
  for (let i = 0; i < 5; i++) {
    const remainder = rating - i;
    ratingArray.push(
      remainder > 1 ? 1 : remainder > 0 ? Number(remainder.toFixed(1)) : 0,
    );
  }

  return ratingArray;
};

export const getPositionRoute = (appPosition) => {
  switch (appPosition) {
    case AppPosition.HOME:
      return "/";
    case AppPosition.SHOP:
      return "/shop";
    case AppPosition.CART:
      return "/cart";
  }
};

export const toTitleCase = (string) => {
  let result = "";
  for (const word of string.split(" ")) {
    result += word[0].toUpperCase() + word.substring(1) + " ";
  }
  return result.trim();
};
