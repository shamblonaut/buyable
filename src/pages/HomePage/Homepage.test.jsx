import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";

import { AppPosition } from "@/utils/constants";

import { mockProductList } from "@tests/data";
import OutletWrapper from "@tests/components/OutletWrapper";

import HomePage from "./HomePage";

const mockOutletContext = {
  productsData: {
    data: mockProductList,
    error: null,
    loading: false,
  },
  appPosition: AppPosition.HOME,
  setAppPosition: vi.spyOn(
    {
      setAppPosition: vi.fn(),
    },
    "setAppPosition",
  ),
};

describe("Home Page", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(
      <OutletWrapper context={mockOutletContext}>
        <HomePage />
      </OutletWrapper>,
    );

    expect(container).toMatchSnapshot();
  });

  it("sets correct global app position", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <HomePage />
      </OutletWrapper>,
    );

    expect(mockOutletContext.setAppPosition).toHaveBeenLastCalledWith(
      AppPosition.HOME,
    );
  });

  it("links to shop page", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <HomePage />
      </OutletWrapper>,
    );

    expect(
      screen.getByRole("link", { name: "Go to Shop page" }),
    ).toHaveAttribute("href", "/shop");
  });
});
