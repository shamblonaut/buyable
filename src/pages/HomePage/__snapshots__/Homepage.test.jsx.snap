// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Page > renders correctly 1`] = `
<div>
  <main
    class="sc-hwkwBN kikxJF"
  >
    <div
      class="sc-eqYatC gWvDHi"
    >
      <ul
        aria-label="carousel-image-list"
        class="sc-kcLKEh kPiXrd"
      >
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/1"
          >
            <img
              alt="Proton L20 32GB RAM 2TB SSD Laptop"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-1.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/2"
          >
            <img
              alt="Tungstenite G8 16GB VRAM Graphics Card"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-2.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/3"
          >
            <img
              alt="Element S12 8GB RAM 256GB ROM Smartphone"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-3.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/4"
          >
            <img
              alt="Electron P11 16-Core Processor"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-4.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/5"
          >
            <img
              alt="Nucleon LGA 1700 Motherboard"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-5.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/6"
          >
            <img
              alt="Newton Sports Sneakers"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-6.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/7"
          >
            <img
              alt="Graviton Travel Bag"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-7.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/8"
          >
            <img
              alt="Torqus Chronograph Wristwatch"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-8.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/9"
          >
            <img
              alt="Pendulus Silver Necklace"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-9.jpg"
            />
          </a>
        </li>
        <li
          class="sc-dntSTA kRvAoC"
        >
          <a
            aria-label="product-link"
            class="sc-kCuUfV dSbdKC"
            data-discover="true"
            href="/product/10"
          >
            <img
              alt="Centrifume Titanium Ring"
              class="sc-dNdcvo kkaztV"
              src="http://example.com/image-10.jpg"
            />
          </a>
        </li>
      </ul>
      <div
        class="sc-kvnevz eEMYVt"
      >
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bUEPuE"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
        <button
          aria-label="carousel-navigation-button"
          class="sc-dFVmKS bBFnAS"
        />
      </div>
    </div>
    <section
      class="sc-hdBJTi hOisCb"
    >
      <p
        class="sc-iIvHqT fwOnxF"
      >
        I can't believe it's not
        <br />
        <strong
          class="sc-jxOwhs jDXtxT"
        >
          BUYABLE
        </strong>
      </p>
      <a
        aria-label="shop-page-link"
        class="sc-lcItFd dCPsRW"
        data-discover="true"
        href="/shop"
      >
        Explore Products
      </a>
    </section>
  </main>
</div>
`;
